File: Assignment 1/main.hs
mergeSort :: Ord a => [a] -> [a]
mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs = merge (mergeSort left) (mergeSort right)
    where
        (left, right) = splitAt (length xs `div` 2) xs

merge :: Ord a => [a] -> [a] -> [a]
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys)
    | x <= y    = x : merge xs (y:ys)
    | otherwise = y : merge (x:xs) ys



quickSort :: Ord a => [a] -> [a]
quickSort [] = []
quickSort (x:xs) =
    let smallerSorted = quickSort [a | a <- xs, a <= x]
        biggerSorted = quickSort [a | a <- xs, a > x]
    in  smallerSorted ++ [x] ++ biggerSorted


main :: IO ()
main = do
    let arr = [38, 27, 43, 3, 9, 82, 10]
    
    -- Testing quick sort
    putStrLn "Quick Sorted array:"
    print (quickSort arr)

    -- Testing merge sort
    putStrLn "Merge Sorted array:"
    print (mergeSort arr)

File: Assignment 1/main.c
#include <stdlib.h>
#include <stdio.h>
#include "quick_sort.c"
#include "merge_sort.c"

int main()
{
    int a[] = {38, 27, 43, 3, 9, 82, 10};
    int n = sizeof(a) / sizeof(a[0]);

    // Testing quick sort
    int quickSortArray[] = {38, 27, 43, 3, 9, 82, 10};
    quick_sort(quickSortArray, n);
    printf("Quick Sorted array: ");
    for (int i = 0; i < n; i++)
    {
        printf("%d ", quickSortArray[i]);
    }
    printf("\n");

    // Testing merge sort
    int mergeSortArray[] = {38, 27, 43, 3, 9, 82, 10};
    merge_sort(mergeSortArray, n);
    printf("Merge Sorted array: ");
    for (int i = 0; i < n; i++)
    {
        printf("%d ", mergeSortArray[i]);
    }
    printf("\n");

    return 0;
}
